Add support for windows container app

vinisoto opened the pull request:
---
**This replaces [the original PR](https://github.com/Azure/azure-cli/pull/7321)**

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

adewaleo commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -69,6 +69,7 @@ def load_arguments(self, _):
         c.argument('name', options_list=['--name', '-n'], help="Name of the new app service plan", completer=None)
         c.argument('sku', arg_type=sku_arg_type)
         c.argument('is_linux', action='store_true', required=False, help='host webapp on Linux worker')
+        c.argument('hyper_v', action='store_true', required=False, help='Preview - host webapp on Windows container')
```
Will a argument / parameter name of is_windows be clearer than hyper_v? Did you test help to see what the parameter shows up as?

I am assuming `hyper_v` is a counterpart to `is_linux`, right?

vinisoto replied:
`hyper-v` is indeed `is_linux` counterpart. `is_windows` will be confusing since non-container web apps are Windows based. We chose a term that differentiates from Linux web apps and from Windows "classic" apps. By the way, we are using the same term in the equivalent PowerShell cmdlet for consistency.

This is what help loos like: 
`--hyper-v                      : Preview - host webapp on Windows container.`

adewaleo replied:
I see, thanks for the clarification.

adewaleo commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -1088,7 +1097,9 @@ def create_app_service_plan(cmd, resource_group_name, name, is_linux, sku='B1',
-    plan_def = AppServicePlan(location=location, tags=tags, sku=sku_def, reserved=(is_linux or None), name=name)
+    plan_def = AppServicePlan(location=location, tags=tags, sku=sku_def,
+                              reserved=(is_linux or None), is_xenon=(hyper_v or None), name=name)
+    print(plan_def)
```
This print statement should not be here. I'm guessing this was left over from testing.

adewaleo commented on src/command_modules/azure-cli-appservice/HISTORY.rst:
```diff
@@ -8,7 +8,11 @@ Release History
 * webapp: add support for listing and restoring deleted apps
+* az functionapp create supports creating a linux consumption plan type with a specific runtime
+* az appserviceplan create now supports creating Windows Containers App Service Plans - Preview
+* az webapp create create now supports creating and managing Windows Containers Apps - Preview
```
"create" is used twice here.
i don't see any history changes referring to `appservice plan update` now allowing for a new argument.
Are there code changes for `webapp create` and `functionapp create`?

vinisoto replied:
There are ineedchanges to `web app create` for example, [this.](https://github.com/Azure/azure-cli/pull/7390/files#diff-35983163c0f2f417dda0c571d4a4538dR91)

`az functionapp create` should not be part of this PR. Will remove it.

We are not changing `appservice plan update` because changing the appservice plan "type" (for example from "Linux" to "Windows" ) is not supported. Once an appservice plan has been created with a certain "type" it cannot be changed.

adewaleo replied:
Thanks for the changes and explanation.
i was mistaken about `appservice plan update`. I thought that it was the command receiving the new `windows_fx_version` parameter.  However, this is a new parameter for the commands `az webapp config set` and `az functionapp config set`.  Shouldn't there be an update to history.rst explaining that this new parameter can be used for these commands? 

Or is this overkill, @yugangw-msft?

adewaleo replied:
I think aside from this last request, this generally looks good to me. Although, @yugangw-msft will also give a look through the PR.

vinisoto replied:
@adewaleo this parameter is intended only for `webapp`s, not `functionapp`s. Although the underlying API won't fail, setting this parameter in a function app doesn't have any effect. I will update the parameter help and updated the end-to-end test to test az webapp config set as suggested.

adewaleo replied:
I see. Thanks for incorporating the changes, @vinisoto 

I would add though that If the parameter is unnecessary for the functionapp command, won't it be better / clearer not to expose it to users at all? One possible way to do this is to use an if statement within the argument context based on the value of scope. 
This way the parameter only shows up for relevant commands and the help text can be more concise / relevant to the command.

What are your thoughts on this?

adewaleo commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -604,7 +613,7 @@ def _get_linux_multicontainer_encoded_config_from_file(file_name):
 # in the method
 def update_site_configs(cmd, resource_group_name, name, slot=None,
-                        linux_fx_version=None, php_version=None, python_version=None,  # pylint: disable=unused-argument
+                        linux_fx_version=None, windows_fx_version=None, php_version=None, python_version=None,  # pylint: disable=unused-argument
```
is this parameter used in this method?

adewaleo replied:
I see it is used in lines 640 - 642.

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -69,6 +69,7 @@ def load_arguments(self, _):
         c.argument('name', options_list=['--name', '-n'], help="Name of the new app service plan", completer=None)
         c.argument('sku', arg_type=sku_arg_type)
         c.argument('is_linux', action='store_true', required=False, help='host webapp on Linux worker')
+        c.argument('hyper_v', action='store_true', required=False, help='Preview - host webapp on Windows container')
```
the convention is `[PREVIEW]....`

adewaleo replied:
Seems like you didn't push your change

vinisoto replied:
Hi @adewaleo . I did. I can see the change when I look to the commit link: https://github.com/Azure/azure-cli/pull/7390/commits/a35af1243c228f3b98c16a47c3f643803a48b84e but somehow it doesn't refresh the main PR page. Let me see what is wrong.

vinisoto replied:
I see what happened. I fixed it in webapp create but missed it in appservice plan create. I fixed it now.

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -73,27 +73,31 @@ def create_webapp(cmd, resource_group_name, name, plan, runtime=None, startup_fi
-    if is_linux:
-        if not validate_linux_create_options(runtime, deployment_container_image_name,
-                                             multicontainer_config_type, multicontainer_config_file):
+    if is_linux or plan_info.is_xenon:
```
i suggest you refactor, say fork out 2 dedicate methods one for linux and the other for windows, then write unit tests to make sure the configuration is right.
the code here is unreadable now with layers of patch

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -557,18 +562,22 @@ def _format_linux_fx_version(custom_image_name, container_config_type=None):
+def _get_fx_version(cmd, resource_group_name, name, slot=None):
     site_config = get_site_configs(cmd, resource_group_name, name, slot)
-    return site_config.linux_fx_version
+    return site_config.linux_fx_version if site_config.linux_fx_version else site_config.windows_fx_version
```
it can be simplified `site_config.linux_fx_version or site_config.windows_fx_version`

vinisoto commented:
Hi @yugangw-msft and @adewaleo - thanks for reviewing. I believe your change requests have been addressed now.

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -1088,7 +1096,8 @@ def create_app_service_plan(cmd, resource_group_name, name, is_linux, sku='B1',
     # the api is odd on parameter naming, have to live with it for now
     sku_def = SkuDescription(tier=get_sku_name(sku), name=sku, capacity=number_of_workers)
-    plan_def = AppServicePlan(location=location, tags=tags, sku=sku_def, reserved=(is_linux or None), name=name)
+    plan_def = AppServicePlan(location=location, tags=tags, sku=sku_def,
```
please error out if both `is_linux` and `hyper_v` are specified.

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -194,6 +195,9 @@ def load_arguments(self, _):
             c.argument('python_version', help='The version used to run your web app if using Python, e.g., 2.7, 3.4')
             c.argument('net_framework_version', help="The version used to run your web app if using .NET Framework, e.g., 'v4.0' for .NET 4.6 and 'v3.0' for .NET 3.5")
             c.argument('linux_fx_version', help="The runtime stack used for your linux-based webapp, e.g., \"RUBY|2.3\", \"NODE|6.6\", \"PHP|5.6\", \"DOTNETCORE|1.1.0\". See https://aka.ms/linux-stacks for more info.")
+            c.argument('windows_fx_version', help="[Preview] Windows Containers web app only. The docker image name used for your windows container webapp, e.g., microsoft/nanoserver:ltsc2016")
```
Suggest remove some redundant wordings: `[Preview] a docker image name used for your Windows container based webapp e.g., microsoft/nanoserver:ltsc2016"`

yugangw-msft commented on src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_params.py:
```diff
@@ -69,6 +69,7 @@ def load_arguments(self, _):
         c.argument('name', options_list=['--name', '-n'], help="Name of the new app service plan", completer=None)
         c.argument('sku', arg_type=sku_arg_type)
         c.argument('is_linux', action='store_true', required=False, help='host webapp on Linux worker')
+        c.argument('hyper_v', action='store_true', required=False, help='[Preview] - host webapp on Windows container')
```
Suggest be consistent `'[Preview] host webapp on Windows container'`

yugangw-msft commented on src/command_modules/azure-cli-appservice/HISTORY.rst:
```diff
@@ -2,6 +2,12 @@
++++++
+* az appserviceplan create now supports creating Windows Containers App Service Plans - Preview
+* az webapp create now supports creating and managing Windows Containers Apps - Preview
+* az webapp config set now supports --windows-fx-version for changing the container image for Windows Containers Apps - Preview
```
suggest simplify the whole 3 lines to: `[preview] support webapps hosting on windows container`

yugangw-msft replied:
in the `_help.py`, add an example under `az webapp create` to make the new things more visible

adewaleo replied:
Thanks for the feedback @yugangw-msft  I was the one who asked that each of these changes be added to history.rst. Your suggested change is more concsise.

vinisoto commented:
Thanks @yugangw-msft - I have addressed the suggestions.  However, looks like [the servicebus module is taking too long to load and is breaking the build](https://travis-ci.org/Azure/azure-cli/jobs/434398142). Any ideas on who/how to address this? I restarted the job a couple of times but keeps failing.

vinisoto commented:
Looks like it passed now.
